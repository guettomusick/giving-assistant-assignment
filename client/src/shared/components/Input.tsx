import React, { FC, useState } from 'react';
import styled from 'styled-components';

import Button from './Button';
import { ChangeEvent } from '../hooks/useHandleChange';

type InputContainerProps = {
  hasTextMuted: boolean,
};

const InputContainer = styled.div<InputContainerProps>`
  position: relative;
  display: flex;
  flex-direction: row;
  width: 100%;

  ${ props => (props.hasTextMuted && `
    padding-bottom: 20px;
  `)};
`;

type StyledInputProps = {
  hasButton: boolean,
};

const StyledInput = styled.input<StyledInputProps>`
  height: 38px;
  border-radius: 6px;
  border 1px solid #666;
  border-shadows: none;
  background: white;
  padding: 8px 14px;
  font-size: 1rem;
  flex-grow: 1;

  ${ props => (props.hasButton && `
    border-radius: 6px 0 0 6px;
    border-right: none;
  `)};
`;

const StyledButton = styled(Button)`
  border-radius: 0 6px 6px 0;
`;

const TextMuted = styled.div`
  position: absolute;
  top: 42px;
  font-size: 0.65em;
  /* text-overflow: revert; */
  color: red;
  width: 100%;
  text-align: center;
`;

export type Props = {
  name: string,
  value: string,
  onChange?: (event: ChangeEvent) => any,
  onClick?: () => any,
  cta?: string,
  textMuted?: string,
  placeholder?: string,
  disabled?: boolean,
  className?: string,
  inputRef?: any,
};

const Input: FC<Props> = ({
  name,
  value,
  onChange,
  onClick,
  cta,
  textMuted,
  placeholder,
  disabled,
  className,
  inputRef,
}) => {
  const [dirty, setDirty] = useState<boolean>(value.length > 0);
  const hasButton = onClick !== undefined && cta !== undefined;
  const hasTextMuted = !!textMuted;

  const handleChange = (event: ChangeEvent) => {
    setDirty(true);
    onChange && onChange(event);
  };

  return (
    <InputContainer className={ className } hasTextMuted={ hasTextMuted }>
      <StyledInput
        name={ name }
        hasButton={ hasButton }
        value={ value }
        onChange={ handleChange }
        placeholder={ placeholder }
        disabled={ disabled }
        ref={ inputRef }
      />
      { hasButton && <StyledButton onClick={ onClick }>{ cta }</StyledButton> }
      { dirty && textMuted && <TextMuted>{ textMuted }</TextMuted> }
    </InputContainer>
  );
};

export default Input;