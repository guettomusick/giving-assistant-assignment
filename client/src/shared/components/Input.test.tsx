import React from 'react';
import renderer from 'react-test-renderer';

import Input, { Props } from './Input';

describe('Input', () => {
  let props: Props;

  beforeEach(() => {
    props = {
      name: 'Some Input',
      onChange: () => {},
      value: 'Some Valu',
    }
  });

  it('renders without button', () => {
    const component = renderer.create(
      <Input { ...props }/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with button', () => {
    props.onClick = () => {};
    props.cta = 'CTA';

    const component = renderer.create(
      <Input { ...props }/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
