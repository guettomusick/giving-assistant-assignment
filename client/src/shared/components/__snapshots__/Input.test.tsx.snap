// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders with button 1`] = `
<div
  className="sc-AxirZ ftCkFX"
>
  <input
    className="sc-AxiKw lbWWnB"
    name="Some Input"
    onChange={[Function]}
    value="Some Valu"
  />
  <button
    className="sc-AxjAm sc-AxhCb gLFHnk"
    onClick={[Function]}
  >
    CTA
  </button>
</div>
`;

exports[`Input renders without button 1`] = `
<div
  className="sc-AxirZ ftCkFX"
>
  <input
    className="sc-AxiKw bTBEhB"
    name="Some Input"
    onChange={[Function]}
    value="Some Valu"
  />
</div>
`;
